### ISSUE TITLE
[java] AvoidInstantiatingObjectsInLoops: False positive - should not flag objects when assigned to lists/arrays

### ISSUE THREAD

**Issue by Fernal73** [pmd/pmd#2207]
**Affects PMD Version:** 
6.20.0
**Rule:**
[AvoidInstantiatingObjectsInLoops](https://pmd.github.io/latest/pmd_rules_java_performance.html#avoidinstantiatingobjectsinloops)
**Description:**
https://stackoverflow.com/questions/17340421/pmd-avoid-instantiating-new-objects-inside-loops
**Code Sample demonstrating the issue**
From StackOverflow post:

```java
import java.awt.Dimension;

public class PMDDemo {
    public static void main(final String[] args) {
        final Dimension[] arr = new Dimension[10];
        for (int i = 0; i < arr.length; i++) {
            arr[i] = new Dimension(i, i); // rule violation here
        }
    }
}
```

Similarly, if object created is added to a list or collection object.

**Running PMD through:** *[CLI | Ant | Maven | Gradle | Designer | Other]*

Similar issues:
* https://github.com/pmd/pmd/issues/1043#issue-316009471 : calls this case a **false negative**
* https://github.com/pmd/pmd/issues/1615 : wants exception for `new java.util.Date()`

**Comment by winhkey** [comment]
yes, I have faced with the same problem.

**Comment by C-Otto** [comment]
Could you please reopen this? I still experience the false positive with 6.3.0, even when using an example adapted from your tests:

```
public Collection<Car> checkCollection() {
        Collection<Car> thisIsACar = new ArrayList<>();
        for(int i = 0; i < 3; ++i) {
            thisIsACar.add(new Car());
        }
        return thisIsACar;
    }

    private static class Car {
    }
```

**Comment by oowekyala** [comment]
@C-Otto It appears this has been fixed in 6.31.0 as a side effect of #3061

**Comment by oowekyala** [comment]
@C-Otto  The only way I can reproduce your issue then is when Collection is unresolved (eg, when an import for `java.util.*` is missing). Maybe it's actually a failure to resolve the type, not a failure in the rule.

Do you know of an earlier PMD version without this bug? Could you make a MWE?

**Comment by C-Otto** [comment]
I had the import for Collection. Here's a small playground project that shows the error:
https://github.com/C-Otto/playground/tree/pmd2207

```
$ ./gradlew pmdMain

> Task :subproject-one:pmdMain FAILED
/home/cotto/IdeaProjects/playground/subproject-one/src/main/java/de/cotto/playground/subprojectone/Hello.java:18:       Avoid instantiating new objects inside loops
```

I don't know of a previous PMD version that does not show the warning.

**Comment by oowekyala** [comment]
Thanks, I can reproduce that. I'll look into it

**Comment by adangel** [comment]
@oowekyala Is this now fixed with #3061 or is this a new false-positive?
Since this issue has been originally fixed with 6.26.0, I'd prefer to create a new issue that we can schedule for a new version and mention in the release notes....

**Comment by oowekyala** [comment]
I forgot to update this sorry. I managed to reproduce the issue with the MWE, and also verify that with PMD 6.31 the violation is gone in the same setup. I'm not sure exactly how this is related to the original issue and originally wanted to investigate further. But I'm short on time, and since the problem appears to be fixed, I'm inclined to just let this be.

**Comment by adangel** [comment]
See #3088

**Comment by Fernal73** [comment]
```java
package streams;

import com.github.javafaker.Faker;
import java.util.ArrayList;
import java.util.List;
import java.util.Random;
import java.util.stream.Collectors;

@SuppressWarnings({"PMD.SystemPrintln",
"PMD.LawOfDemeter"})
public final class ForLoops {
  private static Faker faker = new Faker();
  private static Random random = new Random();
  private static final int MAX_SIBLINGS = 10;
  private static final int PERSON_COUNT = 10_000;

  private ForLoops() {
    throw new IllegalStateException("Private constructor");
  }

  public static void main(String... args) {
    List<Person> persons = new ArrayList<>();
    for (int i = 0; i < PERSON_COUNT; i++) {
      Person[] siblings = getNewSiblings();
      for (int j = 0; j < siblings.length; j++)
        persons.add(siblings[j]);
    }
    System.out.println("Number of persons generated: " + persons.size());
    streaming(persons);
    streamingNames(persons);
    streamingMales(persons);
    streamingCount(persons);
    streamingDistinct(persons);
  }

  private static void streamingCount(List<Person> persons) {
    List<String> result =
        persons.stream()
            .flatMap(
                p -> p.hasSiblings() ? p.getSiblings().stream() : p.asStream())
            .distinct()
            .filter(p -> p.getGender().equals("M"))
            .filter(p -> p.getAge() > 18)
            .filter(p -> p.getAge() <= 65)
            .filter(p -> p.getName() != null)
            .filter(p -> p.getName().startsWith("B"))
            .map(p -> p.getName())
            .collect(Collectors.toList());
    System.out.println("Males starting with B (): " + result.size());
  }

  private static void streamingDistinct(List<Person> persons) {
    List<String> result = persons.stream()
                              .flatMap(p -> p.getSiblings().stream())
                              .distinct()
                              .filter(p -> p.getGender().equals("M"))
                              .filter(p -> p.getAge() > 18)
                              .filter(p -> p.getAge() <= 65)
                              .filter(p -> p.getName() != null)
                              .filter(p -> p.getName().startsWith("B"))
                              .map(p -> p.getName())
                              .collect(Collectors.toList());
    System.out.println("Males (who have siblings) starting with B (): "
                       + result.size());
  }

  private static void streamingMales(List<Person> persons) {
    List<String> result = persons.stream()
                              .filter(p -> p.getGender().equals("M"))
                              .filter(p -> p.getAge() > 18)
                              .filter(p -> p.getAge() <= 65)
                              .filter(p -> p.getName() != null)
                              .filter(p -> p.getName().startsWith("B"))
                              .map(p -> p.getName())
                              .collect(Collectors.toList());
    System.out.println("Male Names starting with B: " + result.size());
  }

  private static void streamingNames(List<Person> persons) {
    List<String> result = persons.stream()
                              .filter(p -> p.getAge() > 18)
                              .filter(p -> p.getAge() <= 65)
                              .filter(p -> p.getName() != null)
                              .filter(p -> p.getName().startsWith("B"))
                              .map(p -> p.getName())
                              .collect(Collectors.toList());
    System.out.println("Names starting with B: " + result.size());
  }

  private static void streaming(List<Person> persons) {
    List<String> result = persons.stream()
                              .filter(p -> p.getAge() > 18)
                              .map(p -> p.getName())
                              .collect(Collectors.toList());
    System.out.println("Age greater than 18: " + result.size());
  }

  private static Person[] getNewSiblings() {
    int count = random.nextInt(MAX_SIBLINGS) + 1;
    Person[] persons = new Person[count];
    int ageMarker = faker.number().numberBetween(0, 150);
    int min = (ageMarker - 10 >= 0) ? ageMarker - 10 : 0;
    int max = (ageMarker + 10 >= 150) ? 150 : ageMarker + 10;
    for (int i = 0; i < count; i++) {
      String name = faker.name().fullName();
      int age = faker.number().numberBetween(min, max);
      String gender = faker.bool().bool() ? "M" : "F";
      Person person = new Person(name, age, gender, new ArrayList<>(count - 1));
      persons[i] = person;
    }
    for (int i = 0; i < count; i++) {
      Person person = persons[i];
      List<Person> siblings = person.getSiblings();
      for (int j = 0; j < count; j++) {
        if (i != j)
          siblings.add(persons[j]);
      }
    }
    return persons;
  }
}



```

Can you reopen this issue? Still flags the instantiation of objects in loop despite assigning them to array.

**Comment by oowekyala** [comment]
@Fernal73 Please minimise your example
